packages/passport-interface/src/api/requestCheckIn.ts

```diff
@@ -0,0 +1,72 @@
+import { EdDSATicketPCD, EdDSATicketPCDPackage } from "@pcd/eddsa-ticket-pcd";
+import { ArgumentTypeName } from "@pcd/pcd-types";
+import { SemaphoreIdentityPCDPackage } from "@pcd/semaphore-identity-pcd";
+import { SemaphoreSignaturePCDPackage } from "@pcd/semaphore-signature-pcd";
+import { Identity } from "@semaphore-protocol/identity";
+import urlJoin from "url-join";
+import {
+  CheckTicketInError,
+  CheckTicketInRequest,
+  CheckTicketInResponseValue,
+  ISSUANCE_STRING
+} from "../RequestTypes";
+import { APIResult } from "./apiResult";
+import { httpPost } from "./makeRequest";
+
+/**
+ * Tries to check the user in. This is called by the PCDpass client when
+ * a 'superuser' of a particular event wants to check in a Devconnect
+ * attendee into the event.
+ *
+ * Never rejects. All information encoded in the resolved response.
+ */
+export async function requestCheckIn(
+  passportServerUrl: string,
+  postBody: CheckTicketInRequest
+): Promise<CheckTicketInResult> {
+  return httpPost<CheckTicketInResult>(
+    urlJoin(passportServerUrl, "/issue/check-in"),
+    {
+      // @todo - here and elsewhere - how can we do better than casting, and actually
```

Comment:
> Agreed that's worth doing, if your goal is to isolate application cod from protocol-related errors. At some point you start worrying about things like version skew between server and client, though maybe that's less of an issue for web clients.

makes sense

> Is a cast in TypeScript an assumption, without any checking?

yes, typescript doesn't do any checking on a cast. it shouldn't even really be called a cast.

Reply:
I've used [Zod](https://zod.dev/) for this in the past.

In simple terms, you define a schema like this:

```
const someRequestSchema = z.object({
  name: z
    .string({
      required_error: 'Name is required',
    })
    .trim()
    .min(1, 'Name cannot be empty'),
  email: z
    .string({
      required_error: 'Email is required',
    })
    .trim()
    .min(1, 'Email cannot be empty')
    .email('Invalid email')
  });
```

You can then do `someRequestSchema.safeParse(input)` and get a return value either confirming that the object conforms to the schema, or providing a list of errors. You can also do:

```
type SomeRequest = z.infer<typeof someRequestSchema>
```

So for any defined schema, you also get a type.

You can also used branded/opaque types, to _enforce_ parsing:
```
const someRequestSchema = z.object(/* as above */).brand("someRequest");
type SomeRequest = z.infer<typeof someRequestSchema>;

function someFunction(req: SomeRequest) { /* function body */}

const req = someRequestSchema.parse({ name: "a name", email: "test@example.com"});
someFunction(req); // works
someFunction({ name: "a name", email: "test@example.com"}); // doesn't work, hasn't been parsed
```

I had been thinking of making a simple PR to do this for feeds, because feed responses can come from untrusted sources. By centralizing logic, this PR might make this harder to do it _just_ for feeds, but I'd be happy to look at it once this has been merged.